[ESQL] New and improved autosuggestion

dej611 opened the pull request:
## Summary

![esql_autosuggest_1](https://github.com/elastic/kibana/assets/924948/5e30cb85-4b72-4234-a4fc-098b8360ed98)

With this PR the autocomplete feature for ESQL is set at the same/better level than the previous one in `main`.
In particular:
* New "shared" grammar should be supported for autosuggest
* the suggestion logic is now based on the AST data structure and the logic is mostly generic`*`
* the suggestion logic is now aware of types when proposing fields, functions and variables
* the suggestion logic has automatic re-suggest on specific types (i.e. `+` or other builtin functions)
* previous tests have been mainly retained to verify feature-parity and in most part extended`**`
* some suggestions are more location-aware than others, and can automatically inject `,` together with the selected argument
* new variables are automatically injected with `=` where it make sense

File hierarchy has changed as well in this PR to accomodate better modularization for different services: ast, validation, autocomplete, hover and signature.

Also various bug fixes have been provided for existing features/services:
* `rename` AST walker had some bugs for partial/incomplete commands
* `rename` validation had some bugs due to the bug above
* `ON` option for `ENRICH` was set as mandatory, now it's set to optional
* `ENRICH` options had several bugs on location extents computation. This was due to the specific grammar definition that didn't make it automatic as other commands to define correct location. It should be fixed now.
* some `AST` functions/options contained empty values which were confusing for services built upon it, and all occurency of this type of bug has been fixed
* fixed various function/command definitions


`*` there are still some edge cases that were required to be handled specifically. But most of the times the logic goes thru the `definition` data structure.
`**` some tests - i.e. suggest open brackets - have been removed as now functions are proposed entirely. For the "closing brackets" suggestion the editor is now configured to automatically open/close them.

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

elasticmachine commented:
Pinging @elastic/kibana-visualizations (Team:Visualizations)

stratoula reviewed (commented):
@dej611 this looks fantastic. I found some cases where autocomplete can be improved:

1. On stats after selecting comma, comma is again suggested instead of a new variable
<img width="1250" alt="image" src="https://github.com/elastic/kibana/assets/17003240/c19453dc-753e-4995-bfad-e77bbad5ddb9">

2. The fourth option here is wrong. (We have a bug in Discover table and list and for some reason we display the column with underscores while it is not correct but I will create an issue). Can we not suggest it? Also the third is correct but unfortunately should be like `avg(bytes)` to work properly. Can we add the `` when the user clicks it?

<img width="1323" alt="image" src="https://github.com/elastic/kibana/assets/17003240/57570f64-5d5d-402f-8005-6103b18dbe34">

3. Not autocomplete related but this is red while it should not. This is a valid syntax
<img width="1061" alt="image" src="https://github.com/elastic/kibana/assets/17003240/bc246d8f-fea3-486f-97f0-36d9a82930df">

4. Eval function arguments suggestions are wrong
<img width="1080" alt="image" src="https://github.com/elastic/kibana/assets/17003240/9c074778-2210-48d5-b6b2-40eed16d1882">

5. Nothing is suggested here :(
<img width="1340" alt="image" src="https://github.com/elastic/kibana/assets/17003240/75589b9f-ad18-42b3-b183-5325ab2a47f1">

stratoula commented:
@dej611 we have this problem also in main but it is more prominent here. It seems that we error out on the capitalized version of the BY operator.

<img width="480" alt="image" src="https://github.com/elastic/kibana/assets/17003240/5d023aed-171c-4cc7-bc8a-994b7d9caf27">

Also not related to autocomplete. We can fix on a separate PR.

ninoslavmiskovic commented:
@dej611 this is great work ! Huge improvement! So nice to e.g get suggestions on index nr 2,3,4 and the meta field .

dej611 commented:
> @dej611 we have this problem also in main but it is more prominent here. It seems that we error out on the capitalized version of the BY operator.

Fixed by b07f1579eceb19e6a0e995eb2084de250b46f6f7 and 819e40ce17634a09445f9b7bd819a0fd983b808b (added unit test as well)

> On stats after selecting comma, comma is again suggested instead of a new variable

Fixed by 985486ffe548a41768e0a1158534bd5ac6be1ba8 (added unit test as well)

> Eval function arguments suggestions are wrong

Fixed by e9a514d24d6b60bc30bbe80e72a128f0a3fe600f (added unit test that checks no consecutive fn suggested)

Solving this issue avoiding duplicates has a side effect the fix also for the tooltip scroll and now the scroll is moved on top to fields again.

> Nothing is suggested here :(

Fixed by 2816c95d7f4eedc51fedbbb317773a12f9f329d3 (added unit tests as well)

> Not autocomplete related but this is red while it should not. This is a valid syntax

Fixed by 304621d37a5c41794b264e6bba063b260b2f52b4 (added unit tests as well)

> The fourth option here is wrong. (We have a bug in Discover table and list and for some reason we display the column with underscores while it is not correct but I will create an issue). Can we not suggest it?

Fixed by d448041895e5d0517ae08c85db58be14a84f6a2c (added unit tests as well)

> Also the third is correct but unfortunately should be like avg(bytes) to work properly. Can we add the `` when the user clicks it?

Fixed by 43d8d849f345ef15acdd74f5dc2520f566db2d7c (added unit tests as well)

stratoula reviewed (approved):
Supercalifragilisticexpialidocious! Bugs fixed, changes LGTM!
